quax/_core.py

```diff
@@ -583,4 +584,38 @@ def _(
     return result
 
 
-# TODO: also register higher-order primitives like `lax.cond_p` etc.
+@register(jax.lax.cond_p)
+def _(index: ArrayLike, *args: Union[ArrayValue, ArrayLike], branches: tuple):
+    def quaxed_jaxpr_out_tree(jaxpr):
+        quax_fn = quaxify(core.jaxpr_as_fun(jaxpr))
+        wrapped_fn, out_tree = api_util.flatten_fun_nokwargs(  # pyright: ignore
+            lu.wrap_init(quax_fn), in_tree
+        )
+        in_avals = tuple([core.raise_to_shaped(core.get_aval(x)) for x in args_leaves])
+        quax_jaxpr = pe.trace_to_jaxpr_dynamic(wrapped_fn, in_avals)[0]
+        return core.ClosedJaxpr(quax_jaxpr, ()), out_tree()
+
+    args_leaves, in_tree = jtu.tree_flatten(args)
+
+    quax_branches_jaxpr: List[Optional[core.ClosedJaxpr]] = [None] * len(branches)
+    quax_jaxpr0, out_tree0 = quaxed_jaxpr_out_tree(branches[0])
```

Comment:
I don't think this complexity is necessary. In particular I'd like to avoid reaching so heavily into JAX internals like this.

I think something like the following would work:
```python
flat_args, tree_args = jtu.tree_flatten(args)
tree_outs = []

new_branches = []
for jaxpr in branches:
    def flat_call(flat_args):
        args = jtu.tree_unflatten(tree_args, flat_args)
        out = quaxify(jax.core.jaxpr_as_fun(jaxpr))(*args)
        flat_out, tree_out = jtu.tree_flatten(out)
        tree_outs.append(tree_out)
        return flat_out

    new_jaxpr = jax.make_jaxpr(flat_call)(flat_args)
    new_branches.append(new_jaxpr)

if any(tree_outs_i != tree_outs[0] for tree_outs_i in tree_outs[1:]):
    raise ValueError(...)

flat_out = jax.lax.cond_p.bind(index, *flat_args, branches=tuple(new_branches)
return jtu.tree_unflatten(tree_outs[0], flat_out)
```

Reply:
I did not get we could just unflatten and flatten while creating the quaxed jaxpr. Much simpler indeed!
Committing with those changes right now